libraries/lib-ipc/internal/BufferedIPCChannel.cpp

```diff
@@ -0,0 +1,158 @@
+#include "BufferedIPCChannel.h"
+
+#include <cassert>
+#include <cstring>
+#include <mutex>
+#include "MemoryX.h"
+#include "socket_guard.h"
+
+BufferedIPCChannel::BufferedIPCChannel()
+{
+   mOutputBuffer.reserve(DefaultOutputBufferCapacity);
+}
+
+BufferedIPCChannel::~BufferedIPCChannel()
+{
+   if(mSocket != INVALID_SOCKET)
+      //interrupts any blocking calls
+      CLOSE_SOCKET(mSocket);
+
+   if(mSendRoutine)
+      mSendRoutine->join();
+   if(mRecvRoutine)
+      mRecvRoutine->join();
+}
+
+void BufferedIPCChannel::Send(const void* bytes, size_t length)
+{
+   std::lock_guard lck(mSocketSync);
+
+   auto offset = mOutputBuffer.size();
+   mOutputBuffer.resize(offset + length);
+   memcpy(mOutputBuffer.data() + offset, bytes, length);
+
+   mSendCondition.notify_one();
+}
+
+void BufferedIPCChannel::StartConversation(SOCKET socket, IPCChannelStatusCallback& callback)
+{
+   assert(socket != INVALID_SOCKET);
+   assert(mSocket == INVALID_SOCKET && !mSendRoutine && !mRecvRoutine);
+   mSocket = socket;
+
+   //create "sending" thread first, it should be blocked
+   //on condition until both IPCChannelStatusCallback::OnConnect
+   //and IPCChannel::Send are called (in that sequence)
+   mSendRoutine = std::make_unique<std::thread>([this]
+   {
+      std::vector<char> secondaryOutputBuffer;
+      secondaryOutputBuffer.reserve(DefaultOutputBufferCapacity);
+      
+      while(true)
+      {
+         std::unique_lock lck(mSocketSync);
+         mSendCondition.wait(lck, [this]{ return !mAlive || !mOutputBuffer.empty(); });
+
+         if(!mAlive)
+         {
+            lck.unlock();
+            return;
+         }
+         
+         std::swap(secondaryOutputBuffer, mOutputBuffer);
+         mOutputBuffer.clear();
+
+         lck.unlock();
+
+         {
+            int offset = 0;
+            while(offset < static_cast<int>(secondaryOutputBuffer.size()))
+            {
+               fd_set writefds, exceptfds;
+               FD_ZERO(&writefds);
+               FD_ZERO(&exceptfds);
+               FD_SET(mSocket, &writefds);
+               FD_SET(mSocket, &exceptfds);
+               auto ret = select(NFDS(mSocket), nullptr, &writefds, &exceptfds, nullptr);
+               if(ret == 1)
+               {
+                  //try to send data...
+                  ret = send(
+                     mSocket,
+                     secondaryOutputBuffer.data() + offset,
+                     static_cast<int>(secondaryOutputBuffer.size()) - offset,
+                     0);
+                  if(ret > 0)
+                     offset += ret;
+                  else if (ret == SOCKET_ERROR)
+                     break;//error
+               }
+               else
+                  break;//error
```

Comment:
If either break is reached, there is no indication of the error to the user.   Should the callback be captured in this lambda too, should `OnDisconnect` be called?  Should the channel shut itself down?

Reply:
`OnDisconnect` is called from receiving thread, there is no need to understand the reason of disconnection right now

Reply:
Are we certain every error seen by a sending thread involves an error in the receiving thread in the other process?  Errors can happen many ways.

It looks like there is some possiblity of sent data being lost but the user of the channel is never informed of it.  Maybe error handling could be more complete.